lib/sources.nix

```diff
@@ -254,6 +254,64 @@ let
       outPath = builtins.path { inherit filter name; path = origSrc; };
     };
 
+  # urlToName : (URL | Path | String) -> String
+  #
+  # Transform a URL (or path, or string) into a clean package name.
+  urlToName = url_:
+    let
+      inherit (builtins) stringLength;
+      url = toString url_;
+      base = baseNameOf (lib.removeSuffix "/" (lib.last (lib.splitString ":" url)));
+      # chop away one git or archive-related extension
+      removeExt = name: let
+          matchExt = builtins.match "(.*)\\.(git|tar|zip|gz|tgz|bz|tbz|bz2|tbz2|lzma|txz|xz|zstd)" name;
+        in if matchExt != null then lib.head matchExt else name;
+      # apply function f to string x while the result shrinks
+      shrink = f: x: let v = f x; in if stringLength v < stringLength x then shrink f v else x;
+    in shrink removeExt base;
+
+  # shortRev : (String | Integer) -> String
+  #
+  # Given a package revision (like "refs/tags/v12.0"), produce a short revision ("12.0").
+  shortRev = rev_:
+    let
+      rev = toString rev_;
+      baseRev = baseNameOf rev;
+      matchHash = builtins.match "[a-f0-9]+" baseRev;
+      matchVer = builtins.match "([A-Za-z]+[-_. ]?)*(v)?([0-9.]+.*)" baseRev;
+    in
+      if matchHash != null then builtins.substring 0 7 baseRev
+      else if matchVer != null then lib.last matchVer
+      else baseRev;
+
+  # repoRevToNameFull : (URL | Path | String) -> (String | Integer | null) -> (String | null) -> String
+  #
+  # See `repoRevToName` below.
+  repoRevToNameFull = repo_: rev_: suffix_:
+    let
+      repo = urlToName repo_;
+      rev = if rev_ != null then "-${shortRev rev_}" else "";
+      suffix = if suffix_ != null then "-${suffix_}" else "";
+    in "${repo}${rev}${suffix}-source";
+
+  # repoRevToName : (Bool | String) -> (URL | Path | String) -> (String | Integer | null) -> String -> String
+  #
+  # Produce derivation.name attribute for a given repository URL/path/name and (optionally) its revision/version tag.
+  #
+  # This is used by fetch(zip|git|FromGitHub|hg|svn|etc) to generate discoverable
+  # /nix/store paths.
+  #
+  # This uses a different implementation depending on the `pretty` argument:
+  #  false  -> name everything as "source"
```

Comment:
To be honest I do not like dual-type attributes. This is a serious footgun in a duck-typed language.

And let's remember Nix is not stellar on the error report department:
https://discourse.nixos.org/t/rant-finding-errors-in-module-configs-after-incompatible-upgrades-is-dreadful/46375

Either is everything a string, say ` format ?  "short", "medium", "long"` or we stick to only two options (and then we can use a Boolean).

Reply:
But it is not checked by Nix but by the `config.nix` and NixOS typing infra, and the errors it generates are pretty nice, e.g. with `nameSourcesPrettily = "error"`:

```
$ nix-instantiate ./default.nix -A xlife
error: A definition for option `nameSourcesPrettily' is not of type `boolean or value "full" (singular enum)'. Definition values:
- In `nixpkgs.config': "error"
(use '--show-trace' to show detailed location information)
```

Reply:
Hum, less bad. 

Nonetheless, semantically it is a clear case for an enumeration with three possible states, instead of a weird Boolean-plus-an-extra.